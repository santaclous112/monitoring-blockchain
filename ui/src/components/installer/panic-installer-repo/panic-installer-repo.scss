@import '../../scss/mixins/centralized-container.scss';
@import '../../scss/mixins/heading.scss';
@import '../../scss/mixins/text.scss';
@import '../../scss/mixins/buttons.scss';
@import '../../scss/mixins/data-table-container';
@import '../../scss/mixins/data-tables.scss';

.panic-installer-repo {

  &__container {
    padding-left: 5%;
    padding-right: 5%;

    p {
      @include text();
    }
  }

  &__heading {
    @include heading-container();
  }

  &__title {
    @include heading-title();
  }

  &__step {
    @include heading-step();
  }

  &__text-with-tooltip{
    @include text-with-tooltip();
  }

  &__add-repo-button {
    @include add-button();
  }

  &__data-table-container {
    @include data-table-container();
  }

  &__crud-data-table {
    @include crud-data-table();
  }
}
