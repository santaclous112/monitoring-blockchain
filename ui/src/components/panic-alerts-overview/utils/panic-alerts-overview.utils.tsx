import {h} from "@stencil/core";
import {Alert} from "../../../interfaces/alerts";
import {SubChain, Source} from "../../../interfaces/chains";
import {
    DataTableHeaderType,
    DataTableRecordType
} from "@simply-vc/uikit/dist/types/types/datatable";
import {SelectOptionType} from "@simply-vc/uikit/dist/types/types/select";
import {HelperAPI} from "../../../utils/helpers";
import {SeverityAPI} from "../../../utils/severity";
import {ALERTS_DATA_TABLE_HEADER} from "../../../utils/constants";

export const AlertsOverviewAPI = {
    getChainFilterOptionsFromChains: getChainFilterOptionsFromChains,
    getSourcesFilterOptions: getSourcesFilterOptions,
    getDataTableJSX: getDataTableJSX,
    validateDateTimeOrder: validateDateTimeOrder,
    formatEventData: formatEventData
}

/**
 * Formats sub-chains to {@link SelectOptionType}.
 * @param subChains array of sub-chains to be formatted.
 * @returns populated {@link SelectOptionType} object.
 */
function getChainFilterOptionsFromChains(subChains: SubChain[]): SelectOptionType {
    return subChains.map(chain => ({label: chain.name, value: chain.name}));
}

/**
 * Formats sources to {@link SelectOptionType}.
 * @param sources array of sources to be formatted.
 * @returns populated {@link SelectOptionType} object.
 */
function getSourcesFilterOptions(sources: Source[]): SelectOptionType {
    return sources.map(source => ({label: source.name, value: source.id}));
}

/**
 * Gets the JSX for the data table.
 * @param alerts array of alerts to be displayed.
 * @returns populated data table as {@link JSX.Element}.
 */
function getDataTableJSX(alerts: Alert[]): JSX.Element {
    let hasAlerts = alerts.length > 0;
    const cols: DataTableHeaderType[] = ALERTS_DATA_TABLE_HEADER;
    const rows: DataTableRecordType = hasAlerts ? getDataTableRecordTypeFromAlerts(alerts) : [];
    hasAlerts = rows.length > 0;

    return <svc-data-table
        key={hasAlerts ? 'data-table-no-alerts' : 'data-table-alerts'}
        cols={cols}
        rows={rows}
        last-clicked-column-index={1}
        ordering={"descending"}
        no-records-message={"There are no alerts to display at this time"}
    />
}

/**
 * Gets {@link DataTableRecordType} from an array of alerts.
 * @param alerts array of alerts.
 * @returns populated {@link DataTableRecordType} object.
 */
function getDataTableRecordTypeFromAlerts(alerts: Alert[]): DataTableRecordType {
    return alerts.map(alert => [
        {
            label: SeverityAPI.getSeverityIcon(alert.severity),
            value: alert.severity
        },
        {
            label: new Date(alert.timestamp * 1000).toLocaleString(),
            value: new Date(alert.timestamp * 1000)
        },
        {label: alert.message, value: alert.message}]);
}

/**
 * Validates the 'from' and 'to' datetime values where the 'from' datetime value
 * should be smaller than or equal to the 'to' datetime value.
 * @param from from datetime value in string format.
 * @param to to datetime value in string format.
 * @returns true if valid, false if invalid.
 */
function validateDateTimeOrder(from: string, to: string): boolean {
    // Return true if any of the date/time values is not set.
    if (from === '' || to === '') {
        return true;
    }

    return HelperAPI.dateTimeStringToTimestamp(from) <= HelperAPI.dateTimeStringToTimestamp(to);
}

/**
 * Formats the event data generated by the svc-filter/svc-select components.
 * @param eventData event data in string format.
 * @returns formatted event data as an array of strings.
 */
function formatEventData(eventData: string): string[] {
    // Split selected options from event data string using commas.
    const selectedOptions: string[] = eventData.split(',');

    // Remove empty string element from array if no options are selected.
    if (selectedOptions.length > 0 && selectedOptions[0] === '') {
        selectedOptions.pop();
    }

    return selectedOptions;
}
