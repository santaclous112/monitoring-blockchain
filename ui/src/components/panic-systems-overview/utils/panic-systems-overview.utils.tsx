import {h} from "@stencil/core";
import {MetricAlert, Metrics} from "../../../interfaces/metrics";
import {
    DataTableHeaderType, DataTableRowsType, DataTableRowCellType
} from "@simply-vc/uikit/dist/types/types/datatable";
import {
    SystemMetricKeys,
    Severity,
    SYSTEM_METRICS_DATA_TABLE_HEADER
} from "../../../utils/constants";

export const SystemsOverviewAPI = {
    getDataTableJSX: getDataTableJSX,
    checkIfAnyNotAvailable: checkIfAnyNotAvailable,
}

/**
 * Gets the JSX for the data table.
 * @param metrics array of metrics.
 * @returns populated data table as {@link JSX.Element}.
 */
function getDataTableJSX(metrics: Metrics[]): JSX.Element {
    const cols: DataTableHeaderType[] = SYSTEM_METRICS_DATA_TABLE_HEADER;
    const rows: DataTableRowsType = getDataTableRowsTypeFromMetrics(metrics);
    let hasMetrics: boolean = metrics.length > 0;

    return <svc-data-table
        key={hasMetrics ? 'data-table-no-alerts' : 'data-table-alerts'}
        cols={cols}
        rows={rows}
        last-clicked-column-index={0}
        ordering={'ascending'}
        no-records-message="There are no metrics to display at this time"
    />
}

/**
 * Gets {@link DataTableRowsType} from an array of metrics.
 * @param metrics array of metrics.
 * @returns populated {@link DataTableRowsType} object.
 */
function getDataTableRowsTypeFromMetrics(metrics: Metrics[]): DataTableRowsType {
    return metrics.map(metric => ({
        cells: [
            {label: metric.name, value: metric.id},
            metricToCellType(metric.metrics.CPU, SystemMetricKeys.SYSTEM_CPU_USAGE, metric.metricAlerts),
            metricToCellType(metric.metrics.RAM, SystemMetricKeys.SYSTEM_RAM_USAGE, metric.metricAlerts),
            metricToCellType(metric.metrics.Storage, SystemMetricKeys.SYSTEM_STORAGE_USAGE, metric.metricAlerts),
            metricToCellType(metric.metrics.ProcessMemory),
            metricToCellType(metric.metrics.ProcessVirtualMemory),
            metricToCellType(metric.metrics.OpenFileDescriptors, SystemMetricKeys.OPEN_FILE_DESCRIPTORS, metric.metricAlerts)
        ],
        id: metric.id
    }));
}

/**
 * Formats metrics to {@link DataTableRowCellType}. If the metric passed is a
 * number, the cell's label is succeeded by a percentage. If the metric key and
 * metric alerts are passed, the css classes field is populated accordingly
 * using {@link checkMetricAlerts}.
 * @param metricValue the value of the metric, can be either a number or a string.
 * @param metricKey the metric key of the passed metric, optional.
 * @param metricAlerts array of metric alerts, optional.
 * @returns populated {@link DataTableRowCellType} object.
 */
function metricToCellType(metricValue: number | string, metricKey?: SystemMetricKeys, metricAlerts?: MetricAlert[]): DataTableRowCellType {
    if (metricValue === undefined || metricValue === null || Number.isNaN(metricValue)) {
        return {label: 'N/A', value: null, cssClasses: ['not-available']};
    }

    const cellType: DataTableRowCellType = {
        label: typeof metricValue === 'number' ? `${metricValue}%` : metricValue,
        value: metricValue
    };

    if (metricKey && metricAlerts) {
        cellType.cssClasses = [checkMetricAlerts(metricKey, metricAlerts)];
    }

    return cellType;
}

/**
 * Returns the status of a given metric based on an array of metric alerts.
 * @param metric metric key being checked.
 * @param metricAlerts array of metric alerts.
 * @returns string with severity of metric or 'INFO' severity if no metric alert
 * is found.
 */
function checkMetricAlerts(metric: SystemMetricKeys, metricAlerts: MetricAlert[]): string {
    const metricAlert: MetricAlert = metricAlerts.find(
      alert => SystemMetricKeys[alert.metric] === metric);

    if (metricAlert !== undefined) {
        return Severity[metricAlert.severity].toLowerCase();
    }

    return Severity.INFO.toLowerCase();
}

/**
 * Checks whether any metric is null (not available).
 * @param metrics array of metrics.
 * @returns true if any metric is null, otherwise returns false.
 */
function checkIfAnyNotAvailable(metrics: Metrics[]): boolean {
    return metrics.some((systemMetric) => {
        return Object.values(systemMetric.metrics).some(
          metric => metric === undefined || metric === null);
    });
}
