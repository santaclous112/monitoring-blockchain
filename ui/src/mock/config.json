[
  {
    "id": "62675e2d8891cd77b87f5b16",
    "status": true,
    "created": "2022-05-02T22:02:46.000Z",
    "modified": null,
    "baseChain": {
      "id": "6265cefcfdb17d641746dced",
      "status": true,
      "name": "Cosmos",
      "value": "cosmos",
      "description": "Cosmos description here...",
      "sources": [
        {
          "id": "6271178cf740ef3847a7d27e",
          "status": true,
          "created": "2022-04-01T00:00:00.000Z",
          "modified": null,
          "name": "cosmos_node_1",
          "value": null,
          "description": null
        }
      ]
    },
    "subChain": {
      "id": "626aeb43980cf43ee7bbf683",
      "status": true,
      "created": "2022-05-02T22:02:46.000Z",
      "modified": null,
      "name": "Akash"
    }
  },
  {
    "id": "62675e2d8891cd77b87f5b166",
    "status": true,
    "created": "2022-05-02T22:02:46.000Z",
    "modified": null,
    "baseChain": {
      "id": "6265cefcfdb17d641746dcedd",
      "status": true,
      "name": "Cosmos",
      "value": "cosmos",
      "description": "Cosmos description here...",
      "sources": [
        {
          "id": "6271178cf740ef3847a7d27e",
          "status": true,
          "created": "2022-04-01T00:00:00.000Z",
          "modified": null,
          "name": "cosmos_node_2",
          "value": null,
          "description": null
        }
      ]
    },
    "subChain": {
      "id": "626aeb43980cf43ee7bbf6833",
      "status": true,
      "created": "2022-05-02T22:02:46.000Z",
      "modified": null,
      "name": "Akash 2"
    }
  }
]