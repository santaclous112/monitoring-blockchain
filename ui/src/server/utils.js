"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Severities = exports.ERR_STATUS = exports.SUCCESS_STATUS = exports.allElementsInListHaveTypeString = exports.getElementsNotInList = exports.allElementsInList = exports.missingValues = exports.toBool = exports.errorJson = exports.resultJson = void 0;
const resultJson = (result) => ({ result });
exports.resultJson = resultJson;
const errorJson = (error) => ({ error });
exports.errorJson = errorJson;
// Transforms a string representing a boolean as a boolean
const toBool = (boolStr) => ['true', 'yes', 'y'].some((element) => boolStr.toLowerCase().includes(element));
exports.toBool = toBool;
// Checks which keys have values which are missing (null, undefined, '') in
// a given object and returns an array of keys having missing values.
const missingValues = (object) => {
    let missingValuesList = [];
    Object.keys(object).forEach((key) => {
        if (!object[key]) {
            missingValuesList.push(key);
        }
    });
    return missingValuesList;
};
exports.missingValues = missingValues;
const allElementsInList = (elements, list) => {
    const resultList = elements.map((element) => {
        return list.includes(element);
    });
    return resultList.every(Boolean);
};
exports.allElementsInList = allElementsInList;
const getElementsNotInList = (elements, list) => {
    return elements.filter(element => !list.includes(element));
};
exports.getElementsNotInList = getElementsNotInList;
const allElementsInListHaveTypeString = (list) => {
    return list.every(item => typeof (item) === "string");
};
exports.allElementsInListHaveTypeString = allElementsInListHaveTypeString;
exports.SUCCESS_STATUS = 200;
exports.ERR_STATUS = 400;
var Severities;
(function (Severities) {
    Severities["INFO"] = "INFO";
    Severities["WARNING"] = "WARNING";
    Severities["CRITICAL"] = "CRITICAL";
    Severities["ERROR"] = "ERROR";
})(Severities = exports.Severities || (exports.Severities = {}));
