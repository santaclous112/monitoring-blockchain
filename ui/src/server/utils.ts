// These functions wrap a result or an error as an object
import Type from "module";

export const resultJson = (result: any) => ({result});
export const errorJson = (error: any) => ({error});

// Transforms a string representing a boolean as a boolean
export const toBool = (boolStr: string): boolean => ['true', 'yes', 'y'].some(
    (element) => boolStr.toLowerCase().includes(element));

// Checks which keys have values which are missing (null, undefined, '') in
// a given object and returns an array of keys having missing values.
export const missingValues = (object: { [id: string]: any }): string[] => {
    let missingValuesList: string[] = [];
    Object.keys(object).forEach((key) => {
        if (!object[key]) {
            missingValuesList.push(key);
        }
    });
    return missingValuesList;
};

export const allElementsInList = (elements: any[], list: any[]): boolean => {
    const resultList: boolean[] = elements.map((element: any): boolean => {
        return list.includes(element)
    });
    return resultList.every(Boolean);
};

export const getElementsNotInList = (elements: any[], list: any[]): any[] => {
    return elements.filter(element => !list.includes(element));
};

export const allElementsInListHaveTypeString = (list: any[]): boolean => {
    return list.every(item => typeof(item) === "string")
};

export const SUCCESS_STATUS: number = 200;
export const ERR_STATUS: number = 400;
