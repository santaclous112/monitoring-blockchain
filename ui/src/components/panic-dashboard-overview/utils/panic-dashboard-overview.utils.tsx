import { h } from '@stencil/core';
import { Alert, Severity } from '../../../interfaces/alerts';
import { BaseChain } from '../../../interfaces/chains';
import { DataTableRecordType } from '../../../lib/types/types/datatable';
import { OrderingType } from '../../../lib/types/types/ordering';
import { SelectOptionType } from '../../../lib/types/types/select';
import { AlertsAPI } from '../../../utils/alerts';

export const DashboardOverviewAPI = {
    getChainFilterOptionsFromBaseChain: getChainFilterOptionsFromBaseChain,
    getPieChartJSX: getPieChartJSX,
    getDataTableJSX: getDataTableJSX
}

/**
 * Formats base chain to SelectOptionType type.
 * @param baseChain base chain to be converted.
 * @returns populated list of required object type.
 */
function getChainFilterOptionsFromBaseChain(baseChain: BaseChain): SelectOptionType {
    return baseChain.chains.map(chain => ({ label: chain.name, value: chain.name }))
}

/**
 * Gets the JSX for the pie chart.
 * @param chainName the respective chain's name.
 * @param criticalAlerts number of critical alerts.
 * @param warningAlerts number of warning alerts.
 * @param errorAlerts number of error alerts.
 * @returns populated pie chart JSX.
 */
function getPieChartJSX(baseChain: BaseChain): JSX.Element {
    let criticalAlerts: number = 0;
    let warningAlerts: number = 0;
    let errorAlerts: number = 0;

    for (const chain of baseChain.chains) {
        if (chain.active) {
            criticalAlerts += chain.alerts.filter(alert => Severity[alert.severity] === Severity.CRITICAL).length;
            warningAlerts += chain.alerts.filter(alert => Severity[alert.severity] === Severity.WARNING).length;
            errorAlerts += chain.alerts.filter(alert => Severity[alert.severity] === Severity.ERROR).length;
        }
    }

    const hasAlerts: boolean = criticalAlerts + warningAlerts + errorAlerts > 0;
    // PieChart config with alerts
    const cols = [{ title: 'Alert', type: 'string' }, { title: 'Amount', type: 'number' }];
    const rows = [['Critical', criticalAlerts], ['Warning', warningAlerts], ['Error', errorAlerts]];
    const alertsColors: string[] = ['#a39293', '#f4dd77', '#f7797b'];
    // PieChart config without alerts
    const noAlertsRows = [['', 1]];
    const noAlertsColors: string[] = ['#b0ea8f'];

    return <svc-pie-chart
        key={hasAlerts ? `${baseChain.name}-pie-chart-no-alerts` : `${baseChain.name}-pie-chart-alerts`}
        colors={hasAlerts ? alertsColors : noAlertsColors}
        cols={cols}
        rows={hasAlerts ? rows : noAlertsRows}
        pie-slice-text={hasAlerts ? "percentage" : "none"}
        tooltip-trigger={hasAlerts ? "focus" : "none"}
    />
}

/**
 * Gets the JSX for the data table.
 * @param alerts list of alerts to be displayed.
 * @returns populated data table JSX.
 */
function getDataTableJSX(baseChain: BaseChain): JSX.Element {
    let alerts: Alert[] = [];

    for (const chain of baseChain.chains) {
        if (chain.active) {
            alerts.push.apply(alerts, chain.alerts);
        }
    }

    const hasAlerts = alerts.length > 0;
    const cols: string[] = ['Severity', 'Time Stamp', 'Message'];
    const rows: DataTableRecordType = hasAlerts ? getDataTableRecordTypeFromAlerts(alerts, baseChain.activeSeverities) : [];

    return <svc-data-table
        id={baseChain.name}
        key={hasAlerts ? `${baseChain.name}-data-table-no-alerts` : `${baseChain.name}-data-table-alerts`}
        cols={cols}
        rows={rows}
        ordering={baseChain.ordering as OrderingType}
        last-clicked-column-index={baseChain.lastClickedColumnIndex}
        no-records-message="There are no alerts to display at this time"
    />
}

/**
 * Filters alerts which do not have an active severity and formats alerts to DataTableRecordType type.
 * @param alerts list of alerts.
 * @param activeSeverities list of active severities.
 * @returns populated list of lists of required object type.
 */
function getDataTableRecordTypeFromAlerts(alerts: Alert[], activeSeverities: Severity[]): DataTableRecordType {
    // Filter alerts.
    const filteredAlerts = alerts.filter(function (alert) {
        return activeSeverities.includes(alert.severity);
    });

    // Format filtered alerts into DataTableRecordType type.
    return filteredAlerts.map(alert => [
        { label: AlertsAPI.getSeverityIcon(alert.severity), value: alert.severity },
        { label: new Date(alert.timestamp * 1000).toLocaleString(), value: new Date(alert.timestamp * 1000) },
        { label: alert.message, value: alert.message }]);
}
