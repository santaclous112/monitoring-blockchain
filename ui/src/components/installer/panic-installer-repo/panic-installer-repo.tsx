import {Component, h, Host, Listen, Prop, State} from '@stencil/core';
import {createAlert, createModal, dismissModal} from "@simply-vc/uikit";
import {HelperAPI} from "../../../utils/helpers";
import {MAIN_TEXT, MORE_INFO_MESSAGES, REPO_TABLE_HEADERS} from "../content/repo";
import {DataTableRowsType} from "@simply-vc/uikit/dist/types/types/datatable";
import {Config} from "../../../../../entities/ts/Config";
import {RepositorySubconfig} from "../../../../../entities/ts/RepositorySubconfig";
import {ConfigService} from "../../../services/config/config.service";
import {RepositoryType} from "../../../../../entities/ts/RepositoryType";
import {Router} from "stencil-router-v2";

@Component({
  tag: 'panic-installer-repo',
  styleUrl: 'panic-installer-repo.scss'
})
export class PanicInstallerRepo {

  /**
   * The identifier used to represent the Config.
   */
  @Prop() configId: string;

  /**
   * Stencil Router object for page navigation.
   */
  @Prop() router: Router;

  /**
   * List containing the Generic Repository Types.
   */
  @Prop() repositoryTypes: RepositoryType[];

  /**
   * The Config object being read & modified.
   */
  @State() config: Config;

  /**
   * Generates a request body to be sent via API.
   * @returns a request object containing the config id and the parsed list of
   * repositories to be sent via the API.
   */
  createRepoRequestBody(): object {
    const repositories = this.config.repositories.map(repository => {
      return {
        id: repository.id,
        name: repository.name,
        namespace: repository.namespace,
        type: repository.type.id,
        monitor: HelperAPI.isTruthy(repository.monitor),
      }
    });

    repositories.forEach(repository=>{
      if(!repository.id){
        delete repository.id;
      }
    })

    return {
      id: this.config.id,
      repositories: repositories,
    };
  }

  /**
   * Save the list of repositories within the config object via the API.
   *
   * @returns the response of the save operation via the API
   */
  async save(): Promise<Response>{
    const requestBody = this.createRepoRequestBody();

    return await ConfigService.getInstance().save(requestBody as Config);
  }

  /**
   * Deletes the repository indicated by the id.
   * @param id the id of the repository to be deleted from in the list
   * @returns the updated RepositorySubconfig[]
   */
  deleteRepositoryFromConfig(id: string): RepositorySubconfig[]{
    return this.config.repositories.filter(repository => repository.id !== id);
  }

  /**
   * Get the repository indicated by the id
   * @param id the id of the repository to be returned
   * @returns RepositorySubconfig that matches the given id
   */
  getRepositoryById(id: string): RepositorySubconfig {
    return this.config.repositories.find((repository) => repository.id === id);
  }

  /**
   * Listens to an event for when a new repo is added/edited via the repo form modal.
   * @param event The event sent by panic-repo-form-modal.
   */
  @Listen("onSave", { target: 'window' })
  async onSaveHandler(event: CustomEvent){
    const repository = event.detail.repositoryObject;
    const id = event.detail.repositoryId;

    //An id indicates that it is an edit scenario. Thus, remove the original repository object from the list.
    //The updated object will be added to this updated list, overall resulting in an update.
    if(id) {
      this.config.repositories = this.deleteRepositoryFromConfig(id);
    }
    this.config.repositories.push(repository);

    const resp = await this.save();

    if(HelperAPI.isDuplicateName(resp)) {
      HelperAPI.raiseToast(
          `A repository named '${repository.name}' already exists!`,
          2000,
          "warning");
      this.config.repositories.pop();
    } else {
      await this.getConfig();
      HelperAPI.raiseToast("Repository added");
      await dismissModal();
    }
  }

  /**
   * Listens to an event sent from the repositories data table when a user clicks an edit button.
   * @param event The event sent by svc-data-table edit functionality.
   */
  @Listen("repositoriesTableEdit", {target: "window"})
  async repositoriesTableEditHandler(event: CustomEvent){
    const repositoryId: string = event.detail.id;
    const repository: RepositorySubconfig = this.getRepositoryById(repositoryId);

    await createModal("panic-repo-form-modal", {
      repositoryTypes: this.repositoryTypes,
      repository: repository,
    }, {backdropDismiss: false});
  }

  /**
   * Listens to an event sent from the repositories data table when a user clicks a delete button.
   * @param event The event sent by svc-data-table deletion functionality.
   */
  @Listen("repositoriesTableDelete", {target: "window"})
  async repositoriesTableDeleteHandler(event: CustomEvent): Promise<void> {
    const repositoryId: string = event.detail.id;
    const repository: RepositorySubconfig = this.getRepositoryById(repositoryId);

    await createAlert({
      header: "Attention",
      message: `Are you sure you want to delete the ${repository.type.name} channel: ${repository.name}?`,
      eventName: 'deleteRepo',
      eventData: {id: repository.id}
    });
  }

  /**
   * Listens to an event sent from the alert created in {@link repositoriesTableDeleteHandler}.
   * @param event The event sent by the delete repository alert.
   */
  @Listen("deleteRepo", { target: 'window' })
  async channelDeleteHandler(event: CustomEvent){
    if(event.detail.confirmed){
      const repositoryId: string = event.detail.data.id;

      this.config.repositories = this.deleteRepositoryFromConfig(repositoryId);

      await this.save();

      await this.getConfig();

      HelperAPI.raiseToast("Repository deleted");
      await dismissModal();
    }
  }

  @Listen("previousStep", {target: "window"})
  previousStepHandler() {
    HelperAPI.changePage(this.router, `${
        HelperAPI.getUrlPrefix()}/sources/${this.configId}`);
  }

  @Listen("nextStep", {target: "window"})
  async nextStepHandler() {
    HelperAPI.changePage(this.router, `${
        HelperAPI.getUrlPrefix()}/alerts/${this.configId}`);
  }

  /**
   * Returns the full name of the repository.
   * For example, in the case of Dockerhub, it returns the namespace/name.
   * Otherwise, returns the name.
   * @param repository the RepositorySubconfig object
   * @returns string
   */
  getRepoFullName(repository: RepositorySubconfig): string {
    if(HelperAPI.isDockerhub(repository)){
      return `${repository.namespace}/${repository.name}`;
    } else {
      return repository.name;
    }
  }

  /**
   * Generates data table rows containing the repository data.
   * @returns DataTableRowsType legible by the svc-data-table.
   */
  repositoryToDataTableRowsType(): DataTableRowsType{
    return this.config.repositories.map(repository => ({
      cells: [
        {
          value: repository.type.value,
          label: repository.type.name,
        },
        {
          value: this.getRepoFullName(repository),
          label: this.getRepoFullName(repository),
        },
      ],
      id: repository.id
    }));
  }

  /**
   * Gets the config object from the API.
   */
  async getConfig() {
    this.config = await ConfigService.getInstance().getByID(this.configId);
  }

  async componentWillLoad() {
    // might be necessary
    // i.e. when the node operator opens a modal and then clicks in the browser's back button
    dismissModal();

    await this.getConfig();
  }

  render() {
    return (
      <Host>
        <panic-header showMenu={false} />

        <svc-progress-bar value={0.8} color={"tertiary"}/>

        <svc-content-container class={"panic-installer-repo__container"}>
          <svc-surface>
            <div class={"panic-installer-repo__heading"}>
              <svc-icon name={"code-slash"} size={"120px"} color={"primary"} />
              <svc-label class={"panic-installer-repo__title"}>Repository Setup for {this.config.subChain.name}</svc-label>
              <svc-label class={"panic-installer-repo__step"}>step 4/5</svc-label>
            </div>

            <div class={"panic-installer-repo__text-with-tooltip"}>
              <p>
                {MAIN_TEXT}
              </p>
              <svc-buttons-container>
                <svc-button color={"secondary"} iconName={"information-circle"} iconPosition={"icon-only"} onClick={async () => {
                  await createModal("panic-installer-more-info-modal", {
                    messages: MORE_INFO_MESSAGES,
                    class: "repo",
                  })
                }}/>
              </svc-buttons-container>
            </div>

            <div class={"panic-installer-repo__add-repo-button"}>
              <svc-button id={"add"} iconName={"add-circle"} color={"secondary"} onClick={async () => {
                await createModal("panic-repo-form-modal", {
                  repositoryTypes: this.repositoryTypes
                }, {backdropDismiss: false});
              }}>
                Add Repository
              </svc-button>
            </div>

            {
              this.config.repositories && this.config.repositories.length > 0 &&
                <div class={"panic-installer-repo__crud-data-table"}>
                  <svc-data-table
                    class={"panic-installer-repo__action_icons"}
                    mode={"crud"}
                    cols={REPO_TABLE_HEADERS}
                    noRecordsMessage={"Your configured repositories will show here."}
                    rows={this.repositoryToDataTableRowsType()}
                    editEventName={"repositoriesTableEdit"}
                    deleteEventName={"repositoriesTableDelete"}
                  />
                </div>
            }
          </svc-surface>

          <panic-installer-nav config={this.config} />

        </svc-content-container>

        <panic-footer />
      </Host>
    );
  }
}
