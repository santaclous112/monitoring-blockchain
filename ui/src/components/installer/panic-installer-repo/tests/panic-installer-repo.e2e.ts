import {newSpecPage} from "@stencil/core/testing";
import {PanicInstallerRepo} from "../panic-installer-repo";
import {ConfigService} from "../../../../services/config/config.service";
import {mockGetConfigData} from "../../../../utils/mock";

beforeAll(() => {
  jest.spyOn(ConfigService.prototype, 'getByID').mockImplementation(
    (_: string) => mockGetConfigData()
  );
});

afterAll(() => {
  jest.restoreAllMocks();
});

describe('panic-installer-repo', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [PanicInstallerRepo],
      html: '<panic-installer-repo></panic-installer-repo>',
    });

    const PanicInstallerRepos = page.body.querySelector('panic-installer-repo');

    expect(PanicInstallerRepos).toMatchSnapshot();
  });
})