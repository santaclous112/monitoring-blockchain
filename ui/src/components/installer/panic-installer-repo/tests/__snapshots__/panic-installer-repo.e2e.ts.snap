// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`panic-installer-repo renders 1`] = `
<panic-installer-repo>
  <panic-header></panic-header>
  <svc-progress-bar color="tertiary" value="0.8"></svc-progress-bar>
  <svc-content-container class="panic-installer-repo__container">
    <svc-surface>
      <div class="panic-installer-repo__heading">
        <svc-icon color="primary" name="code-slash" size="120px"></svc-icon>
        <svc-label class="panic-installer-repo__title">
          Repository Setup for test_sub_chain
        </svc-label>
        <svc-label class="panic-installer-repo__step">
          step 4/5
        </svc-label>
      </div>
      <div class="panic-installer-repo__text-with-tooltip">
        <p>
          PANIC allows you to monitor GitHub and DockerHub repositories for new releases/tags respectively.
        </p>
        <svc-buttons-container>
          <svc-button color="secondary" iconname="information-circle" iconposition="icon-only"></svc-button>
        </svc-buttons-container>
      </div>
      <div class="panic-installer-repo__add-repo-button">
        <svc-button color="secondary" iconname="add-circle" id="add">
          Add Repository
        </svc-button>
      </div>
    </svc-surface>
    <panic-installer-nav></panic-installer-nav>
  </svc-content-container>
  <panic-footer></panic-footer>
</panic-installer-repo>
`;
